libtenzir/builtins/operators/unordered.cpp

```diff
@@ -0,0 +1,103 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/parser_interface.hpp>
+#include <tenzir/pipeline.hpp>
+#include <tenzir/plugin.hpp>
+
+#include <arrow/type.h>
+
+namespace tenzir::plugins::unordered {
+
+namespace {
+
+class unordered_operator final : public operator_base {
+public:
+  unordered_operator() = default;
+
+  explicit unordered_operator(operator_ptr op) : op_{std::move(op)} {
+    if (auto* op = dynamic_cast<unordered_operator*>(op_.get())) {
+      op_ = std::move(op->op_);
+    }
+    TENZIR_ASSERT(not dynamic_cast<const unordered_operator*>(op_.get()));
+  }
+
+  auto optimize(const expression& filter, event_order order) const
+    -> optimize_result override {
+    (void)order;
+    return op_->optimize(filter, event_order::unordered);
+  }
```

Comment:
The `event_order::unordered` argument allows `op` to assume that the behavior of the pipeline that comes after it does not depend on the order of its input. I think you want to modify the returned `optimize_result::order` instead perhaps? Or perhaps even both?